docs/source/config/settings.rst
Settings
========

``indico.conf`` is Indico's main configuration file. Its initial version
is usually generated when running ``indico setup wizard`` as described in
the Installation Guide, but depending on the setup it should be modified
later.

The config file is loaded from the path specified in the ``INDICO_CONFIG``
environment variable; if no such path is set, the config file (or a symlink
to it) is searched in the following places, in order:

- ``<indico_package_path>/indico.conf`` (development setups only)
- ``~/.indico.conf``
- ``/etc/indico.conf``

The file is executed as a Python module, so anything that is valid Python
2.7 code can be used in it. When defining temporary variables that are not
config options, their name should be prefixed with an underscore; otherwise
you will get a warning about unknowing config options being defined.


.. _settings-auth:

Authentication
--------------

.. data:: LOCAL_IDENTITIES

    This setting controls whether local Indico accounts are available.
    If no centralized authentication infrastructure (e.g. LDAP, OAuth,
    or another kind of SSO) is used, local accounts are the only way
    of logging in to Indico.

    Default: ``True``

.. data:: LOCAL_GROUPS

    This setting controls whether local Indico groups are available.
    If no centralized authentication infrastructure that supports groups
    (e.g. LDAP) is used, local groups are the only way to define groups in
    Indico, but if you do have central groups it may be useful to disable
    local ones to have all groups in one central place.

    Default: ``True``

.. data:: LOCAL_REGISTRATION

    This setting controls whether people accessing Indico can create a
    new account.  Admins can always create new local accounts, regardless
    of this setting.

    This setting is only taken into account if :data:`LOCAL_IDENTITIES`
    are enabled.

    Default: ``True``

.. data:: LOCAL_MODERATION

    This setting controls whether a new registration needs to be approved
    by an admin before the account is actually created.

    This setting is only taken into account if :data:`LOCAL_IDENTITIES`
    and :data:`LOCAL_REGISTRATION` are enabled.

    Default: ``False``

.. data:: EXTERNAL_REGISTRATION_URL

    The URL to an external page where people can register an account that
    can then be used to login to Indico (usually via LDAP/SSO).

    This setting is only taken into account if :data:`LOCAL_IDENTITIES`
    are disabled.

    Default: ``None``

.. data:: AUTH_PROVIDERS

    A dict defining `Flask-Multipass`_ authentication providers used
    by Indico.  The dict specified here is passed to the
    ``MULTIPASS_AUTH_PROVIDERS`` setting of Flask-Multipass.

    Default: ``{}``

.. data:: IDENTITY_PROVIDERS

    A dict defining `Flask-Multipass`_ identity providers used by Indico
    to look up user information based on the data provided by an
    authentication provider.  The dict specified here is passed to the
    ``MULTIPASS_IDENTITY_PROVIDERS`` setting of Flask-Multipass.

    Default: ``{}``

.. data:: PROVIDER_MAP

    If not specified, authentication and identity providers with the
    same name are linked automatically.  The dict specified here is
    passed to the ``MULTIPASS_PROVIDER_MAP`` setting of Flask-Multipass.

    Default: ``{}``


Cache
-----

.. data:: CACHE_BACKEND

    The backend used for caching. Valid backends are ``redis``,
    ``files``, and ``memcached``.

    To use the ``redis`` backend (recommended), you need to set
    :data:`REDIS_CACHE_URL` to the URL of your Redis instance.

    With the ``files`` backend, cache data is stored in :data:`CACHE_DIR`,
    which always needs to be set, even when using a different cache
    backend since Indico needs to cache some data on disk.

    To use the ``memcached`` backend, you need to install the
    ``python-memcached`` package from PyPI and set :data:`MEMCACHED_SERVERS`
    to a list containing at least one memcached server.

    .. note::

        We only test Indico with the ``redis`` cache backend. While
        the other backends should work, we make no guarantees as
        they are not actively being used or tested.

    Default: ``'files'``

.. data:: REDIS_CACHE_URL

    The URL of the redis server to use with the ``redis`` cache backend.

    If the Redis server requires authentication, use a URL like this:
    ``redis://unused:password@127.0.0.1:6379/1``

    If no authentication is used (usually the case with a local Redis
    server), you can omit the user/password part:
    ``redis://127.0.0.1:6379/1``

    Default: ``None``

.. data:: MEMCACHED_SERVERS

    The list of memcached servers (each entry is an ``ip:port`` string)
    to use with the ``memcached`` cache backend.

    Default: ``[]``


Celery
------

.. data:: CELERY_BROKER

    The URL of the Celery broker (usually Redis of AMQP) used for
    communication between Indico and the Celery background workers.

    We recommend using Redis as it is the easiest option, but you can
    check the `Celery documentation on brokers`_ for more information
    on the other possible brokers.

    Default: ``None``

.. data:: CELERY_RESULT_BACKEND

    The URL of the Celery result backend. If not set, the same backend
    as the broker is used.  Indico currently does not use task results,
    and we recommend leaving this setting at its default.

    Default: ``None``

.. data:: CELERY_CONFIG

    A dict containing additional Celery settings.

    .. warning::

        This is an advanced setting that is rarely needed and we do not
        recommend using it unless you know exactly what you are doing!
        Changing Celery settings may break things or result in tasks not
        being executed without other changes (such as running additional
        celery workers on different queues).

    One use case for this setting is routing certain tasks to a different
    queue, and then running multiple Celery workers for these queues.

    .. code-block:: python

        CELERY_CONFIG = {
            'task_routes': {
                'indico_livesync.task.scheduled_update': {'queue': 'livesync'},
            }
        }

    Default: ``{}``

.. data:: SCHEDULED_TASK_OVERRIDE

    A dict overriding the task schedule for specific tasks.

    By default, all periodic tasks are enabled and use a schedule which
    we consider useful for most cases.  Using this setting, you can
    override the default schedule.

    The dict key is the name of the task and the value can be one of
    the following:

    - ``None`` or ``False`` -- disables the task completely
    - A dictionary, as described in the `Celery documentation on periodic tasks`_.
      The ``task`` should not be specified, as it is set automatically.
    - A :class:`~datetime.timedelta` or :class:`~celery.schedules.crontab`
      object which will just override the schedule without changing any
      other options of the task.  Both classes are available in the config
      file by default.

    .. note::

        Use ``indico celery inspect registered`` to get a list of task
        names.  Celery must be running for this command to work.

    Default: ``{}``


Customization
-------------

.. data:: CUSTOMIZATION_DIR

    The base path to the directory containing customizations for your
    Indico instance.

    It is possible to override specific templates and add CSS and
    JavaScript for advanced customizations.  When using this, be
    advised that depending on the modifications you perform things
    may break after an Indico update.  Make sure to test all your
    modifications whenever you update Indico!

    To include custom CSS and JavaScript, simply put ``*.css`` and
    ``*.js`` files into ``<CUSTOMIZATION_DIR>/css`` / ``<CUSTOMIZATION_DIR>/js``.
    If there are multiple files, they will be included in alphabetical
    order, so prefixing them with a number (e.g. ``00-base.css``, ``10-events.css``)
    is a good idea.

    Static files may be added in ``<CUSTOMIZATION_DIR>/files``.  They can be
    referenced in templates through the ``assets.custom`` endpoint.  In CSS/JS,
    the URL for them needs to be built manually (``/static/custom/files/...``).

    For template customizations, see the description of :data:`CUSTOMIZATION_DEBUG`
    as this setting is highly recommended to figure out where exactly to
    put customized templates.

    Here is an example for a template customization that includes a
    custom asset and uses inheritance to avoid having to replace the
    whole template:

    .. code-block:: jinja

        {% extends '~footer.html' %}

        {% block footer_logo %}
            {%- set filename = 'cern_small_light.png' if dark else 'cern_small.png' -%}
            <a href="https://home.cern/" class="footer-logo">
                <img src="{{ url_for('assets.custom', filename=filename) }}" alt="CERN">
            </a>
        {% endblock %}

    Default: ``None``

.. data:: CUSTOMIZATION_DEBUG

    Whether to log details for all customizable templates the first time
    they are accessed.  The log message contains the path where you need
    to store the template; this path is relative to
    ``<CUSTOMIZATION_DIR>/templates/``.

    The log message also contains the full path of the original template
    in case you decide to copy it.
    However, instead of copying templates it is better to use Jinja
    inheritance where possible.  To make this easier the log entry contains
    a "reference" path that can be used to reference the original template
    from the customized one.

    Default: ``False``

.. data:: HELP_URL

    The URL used for the "Help" link in the footer.

    Default: ``'https://learn.getindico.io'``

.. data:: LOGO_URL

    The URL to a custom logo.  If unset, the default Indico logo is used.

    Default: ``None``

.. data:: CUSTOM_COUNTRIES

    A dict with country name overrides.  This can be useful if the official
    ISO name of a country does not match what your Indico instance's target
    audience expects for a country, e.g. due to political situations.

    .. code-block:: python

        CUSTOM_COUNTRIES = {'KP': 'North Korea'}

    Default: ``{}``

.. data:: CUSTOM_LANGUAGES

    A dict with language/territory name overrides.  This can be useful if the
    official territory name that goes along with a language does not match what
    your Indico instance's target audience expects for a country, e.g. due to
    political situations.

    For example, to replace "Chinese (Simplified)" with "Chinese (China)",
    you would use:

    .. code-block:: python

        CUSTOM_LANGUAGES = {'zh_Hans_CN': ('Chinese', 'Simplified')}

    Note that the language and territory name should be written in that
    particular language to be consistent with the defaults. So in the example
    above, you would write "Chinese" and "Simplified" in Simplified Chinese.

    Setting the territory (second element in the tuple) to ``None`` will hide
    it and only show the language name itself.  Setting the dict value to ``None``
    will effectively hide the language altogether.

    Default: ``{}``


Database
--------

.. data:: SQLALCHEMY_DATABASE_URI

    The URI used to connect to the PostgreSQL database.  For a local database,
    you can usually omit everything besides the database name:
    ``postgresql:///indico``

    If the database requires authentication and/or runs on a separate host,
    this form should be used: ``postgresql://user:password@hostname/dbname``

.. data:: SQLALCHEMY_POOL_SIZE

    This setting configures SQLAlchemy's connection pool.
    For details, check the `Flask-SQLAlchemy documentation`_.

    Default: ``5``

.. data:: SQLALCHEMY_POOL_RECYCLE

    This setting configures SQLAlchemy's connection pool.
    For details, check the `Flask-SQLAlchemy documentation`_.

    Default: ``120``

.. data:: SQLALCHEMY_POOL_TIMEOUT

    This setting configures SQLAlchemy's connection pool.
    For details, check the `Flask-SQLAlchemy documentation`_.

    Default: ``10``


Development
-----------

.. warning::

    Do not turn on development settings in production.  While we are not
    aware of serious security issues caused by these settings, they may
    slow down Indico or remove redundancies and thus make Indico not as
    stable as one would expect it to be in a production environment.

.. data:: DEBUG

    Enables debugging mode.  If enabled, assets are not minified, error
    messages are more verbose and various other features are configured
    in a developer-friendly way.

    **Do not enable debug mode in production.**

    Default: ``False``

.. data:: DB_LOG

    Enables real-time database query logging.  When enabled, all database
    queries are sent to a socket where they can be read by the ``db_log.py``
    script.  To use the database logger, run ``bin/utils/db_log.py`` (only
    available when running Indico from a Git clone) in a separate terminal
    and all requests and verbose queries will be displayed there.

    Default: ``False``

.. data:: PROFILE

    Enables the Python profiler.  The profiler output is stored in
    ``<TEMP_DIR>/*.prof``.

    Default: ``False``

.. data:: SMTP_USE_CELERY

    If disabled, emails will be sent immediately instead of being
    handed to a Celery background worker.  This is often more convenient
    during development as you do not need to run a Celery worker while still
    receiving emails sent from Indico.
    Disabling it may result in emails not being sent if the mail server is
    unavailable or some other failure happens during email sending.  Because
    of this, the setting should never be disabled in a production environment.

    Default: ``True``

.. data:: COMMUNITY_HUB_URL

    The URL of the community hub. This should only be changed when using a local
    instance of Mereswine to debug the interface between Indico and Mereswine.

    Default: ``'https://hub.getindico.io'``

.. data:: DISABLE_CELERY_CHECK

    Disables the warning about Celery not running or being outdated. When set to
    ``None``, the warning is disabled when :data:`DEBUG` is enabled; otherwise
    this setting enables/disables the warning regardless of debug mode.

    Default: ``None``


Directories
-----------

.. data:: CACHE_DIR

    The directory in which various data is cached temporarily. Must be
    accessible by the web server.

    Default: ``'/opt/indico/cache'``

.. data:: LOG_DIR

    The directory in which log files are stored. Can be overridden by
    using absolute paths in ``logging.yaml``.

    Default: ``'/opt/indico/log'``

.. data:: TEMP_DIR

    The directory in which various temporary files are stored. Must be
    accessible by the web server.

    Default: ``'/opt/indico/cache'``


Emails
------

.. data:: SMTP_SERVER

    The hostname and port of the SMTP server used for sending emails.

    Default: ``('localhost', 25)``

.. data:: SMTP_LOGIN

    The username to send if the SMTP server requires authentication.

    Default: ``None``

.. data:: SMTP_PASSWORD

    The password to send if the SMTP server requires authentication.

    Default: ``None``

.. data:: SMTP_USE_TLS

    If enabled, STARTTLS will be used to use an encrypted SMTP connection.

    Default: ``False``

.. data:: SMTP_TIMEOUT

    The timeout in seconds after which a connection attempt to the SMTP
    server is aborted.

    Default: ``30``

.. data:: NO_REPLY_EMAIL

    The email address used when sending emails to users to which they
    should not reply.

    Default: ``None``

.. data:: PUBLIC_SUPPORT_EMAIL

    The email address that is shown to users on the "Contact" page.

    Default: ``None``

.. data:: SUPPORT_EMAIL

    The email address of the technical manager of the Indico instance.
    Emails about unhandled errors/exceptions are sent to this address.

    Default: ``None``


Experimental Features
----------------------

.. data:: EXPERIMENTAL_EDITING_SERVICE

    If enabled, event managers can connect the Editing module of their
    events to an external microservice extending the normal Editing workflow.
    As long as this is considered experimental, there are no guarantees
    on backwards compatibility even in minor Indico version bumps. Please
    check the `reference implementation`_ for details/changes.

    Default: ``False``


LaTeX
-----

.. data:: XELATEX_PATH

    The full path to the ``xelatex`` program of `TeXLive`_.

    If it is installed in a directory in your ``$PATH``, specifying its
    name without a path is sufficient.

    If the path is not configured, any functionality that requires LaTeX
    on the server (such as generating the Book of Abstracts or exporting
    contributions to PDF) will be disabled.

    Default: ``None``

.. data:: STRICT_LATEX

    Enables strict mode for LaTeX rendering, in which case a non-zero
    status code is considered failure.

    LaTeX is rather generous when it comes to using a non-zero exit code.
    For example, having an oversized image in an abstract is enough to
    cause one.  It is generally not a good idea to enable strict mode as
    this will result in PDF generation to fail instead of creating a PDF
    that looks slightly uglier (e.g. a truncated image) than one that would
    succeed without a non-zero status code.

    Default: ``False``


Logging
-------

.. data:: LOGGING_CONFIG_FILE

    The path to the logging config file.  Unless an absolute path is specified,
    the path is relative to the location of the Indico config file after
    resolving symlinks.

    Default: ``'logging.yaml'``

.. data:: SENTRY_DSN

    If you use `Sentry`_ for logging warnings/errors, you can specify the
    connection string here.

    Default: ``None``

.. data:: SENTRY_LOGGING_LEVEL

    The minimum level a log record needs to have to be sent to Sentry.
    If you do not care about warnings, set this to ``'ERROR'``.

    Default: ``'WARNING'``


Security
--------

.. data:: SECRET_KEY

    The secret key used to sign tokens in URLs.  It must be kept secret
    under all circumstances.

    When using Indico on a cluster of more than one worker, all machines
    need to have the same secret key.

    The initial key is generated by the setup wizard, but if you have to
    regenerate it, the best way of doing so is running this snippet on a
    shell:  ``python -c 'import os; print repr(os.urandom(32))'``

    Default: ``None``

.. data:: SESSION_LIFETIME

    The duration of inactivity after which a session and its session cookie
    expires.  If set to ``0``, the session cookie will be cleared when the
    browser is closed.

    Default: ``86400 * 31``


Storage
-------

.. data:: STORAGE_BACKENDS

    The list of backends that can be used to store/retrieve files.

    Indico needs to store various files such as event attachments somewhere.
    By default only a filesystem based storage backend is available, but
    plugins could add additional backends.  You can define multiple backends,
    but once a backend has been used, you **MUST NOT** remove it or all
    files stored in that backend will become unavailable.

    To define a filesystem-based backend, use the string ``fs:/base/path``.
    If you stopped using a backend, you can switch it to read-only mode by
    using ``fs-readonly:`` instead of ``fs:``

    Other backends may accept different options - see the documentation of these
    backends for details.

    Default: ``{'default': 'fs:/opt/indico/archive'}``

.. data:: ATTACHMENT_STORAGE

    The name of the storage backend used to store all kinds of attachments.
    Anything in this backend is write-once, i.e. once stored, files in it
    are never modified or deleted.

    Changing this only affects new uploads; existing files are taken from
    the backend that was active when they were uploaded -- which is also
    why you must not remove a backend from :data:`STORAGE_BACKENDS` once
    it has been used.

    Default: ``'default'``

.. data:: STATIC_SITE_STORAGE

    The name of the storage backend used to store "offline copies" of
    events.  Files are written to this backend when generating an offline

CHANGES.rst
Changelog
=========


Version 3.0
-----------

*Unreleased*

Major Features
^^^^^^^^^^^^^^

- TODO

Improvements
^^^^^^^^^^^^

- TODO

Internal Changes
^^^^^^^^^^^^^^^^

- Require Python 3.9 - older Python versions (especially Python 2.7) are **no longer supported**
- The integration with flower (celery monitoring tool) has been removed as it was not widely used,
  did not provide much benefit, and it is no longer compatible with the latest Celery version


----

Version 2.3.2
-------------

*Unreleased*

Improvements
^^^^^^^^^^^^

- Disable title field by default in new registration forms (:issue:`4688`, :pr:`4692`)
- Add gender-neutral "Mx" title (:issue:`4688`, :pr:`4692`)

Bugfixes
^^^^^^^^

- None so far :)

Version 2.3.1
-------------

*Released on October 27, 2020*

Security fixes
^^^^^^^^^^^^^^
- Fix potential data leakage between OAuth-authenticated and unauthenticated HTTP API requests
  for the same resource (:pr:`4663`)

.. note::

    Due to OAuth access to the HTTP API having been broken until this version, we do not
    believe this was actually exploitable on any Indico instance. In addition, only Indico
    administrators can create OAuth applications, so regardless of the bug there is no risk
    for any instance which does not have OAuth applications with the ``read:legacy_api``
    scope.

Improvements
^^^^^^^^^^^^

- Generate material packages in a background task to avoid timeouts or using excessive
  amounts of disk space in case of people submitting several times (:pr:`4630`)
- Add new :data:`EXPERIMENTAL_EDITING_SERVICE` setting to enable extending an event's Editing
  workflow through an `OpenReferee server <https://github.com/indico/openreferee/>`_ (:pr:`4659`)

Bugfixes
^^^^^^^^

- Only show the warning about draft mode in a conference if it actually has any
  contributions or timetable entries
- Do not show incorrect modification deadline in abstract management area if no
  such deadline has been set (:pr:`4650`)
- Fix layout problem when minutes contain overly large embedded images (:issue:`4653`,
  :pr:`4654`)
- Prevent pending registrations from being marked as checked-in (:pr:`4646`, thanks
  :user:`omegak`)
- Fix OAuth access to HTTP API (:pr:`4663`)
- Fix ICS export of events with draft timetable and contribution detail level
  (:pr:`4666`)
- Fix paper revision submission field being displayed for judges/reviewers (:pr:`4667`)
- Fix managers not being able to submit paper revisions on behalf of the user (:pr:`4667`)

Internal Changes
^^^^^^^^^^^^^^^^

- Add ``registration_form_wtform_created`` signal and send form data in
  ``registration_created`` and ``registration_updated`` signals (:pr:`4642`,
  thanks :user:`omegak`)
- Add ``logged_in`` signal


Version 2.3
-----------

*Released on September 14, 2020*

.. note::

    We also published a `blog post <https://getindico.io/indico/update/release/milestone/2020/07/22/indico-2-3-news.html>`_
    summarizing the most relevant changes for end users.

Major Features
^^^^^^^^^^^^^^

- Add category roles, which are similar to local groups but within the
  scope of a category and its subcategories. They can be used for assigning
  permissions in any of these categories and events within such categories.
- Events marked as "Invisible" are now hidden from the category's event list
  for everyone except managers (:issue:`4419`, thanks :user:`openprojects`)
- Introduce profile picture, which is for now only visible on the user dashboard
  (:issue:`4431`, thanks :user:`omegak`)
- Registrants can now be added to event ACLs. This can be used to easily restrict
  parts of an event to registered participants. If registration is open and a registration
  form is in the ACL, people will be able to access the registration form even if they
  would otherwise not have access to the event itself. It is also possible to restrict
  individual event materials and custom page/link menu items to registered participants.
  (:issue:`4477`, :issue:`4528`, :issue:`4505`, :issue:`4507`)
- Add a new Editing module for papers, slides and posters which provides a workflow
  for having a team review the layout/formatting of such proceedings and then publish
  the final version on the page of the corresponding contribution. The Editing module
  can also be connected to an external microservice to handle more advanced workflows
  beyond what is supported natively by Indico.

Internationalization
^^^^^^^^^^^^^^^^^^^^

- New translation: Chinese (Simplified)

Improvements
^^^^^^^^^^^^

- Sort survey list by title (:issue:`3802`)
- Hide "External IDs" field if none are defined (:issue:`3857`)
- Add LaTeX source export for book of abstracts (:issue:`4035`,
  thanks :user:`bpedersen2`)
- Tracks can now be categorized in track groups (:issue:`4052`)
- Program codes for sessions, session blocks, contributions and
  subcontributions can now be auto-generated (:issue:`4026`)
- Add draft mode for the contribution list of conference events
  which hides pages like the contribution list and timetable until
  the event organizers publish the contribution list. (:issue:`4095`)
- Add ICS export for information in the user dashboard (:issue:`4057`)
- Allow data syncing with multipass providers which do not support
  refreshing identity information
- Show more verbose error when email validation fails during event
  registration (:issue:`4177`)
- Add link to external map in room details view (:issue:`4146`)
- Allow up to 9 digits (instead of 6) before the decimal point in
  registration fees
- Add button to booking details modal to copy direct link (:issue:`4230`)
- Do not require new room manager approval when simply shortening a booking
  (:issue:`4214`)
- Make root category description/title customizable using the normal
  category settings form (:issue:`4231`)
- Added new :data:`LOCAL_GROUPS` setting that can be used to fully disable
  local groups (:issue:`4260`)
- Log bulk event category changes in the event log (:issue:`4241`)

indico/core/config.py
# This file is part of Indico.
# Copyright (C) 2002 - 2020 CERN
#
# Indico is free software; you can redistribute it and/or
# modify it under the terms of the MIT License; see the
# LICENSE file for more details.

import ast
import codecs
import os
import re
import socket
import warnings
from datetime import timedelta

import pytz
from celery.schedules import crontab
from flask import current_app, g
from flask.helpers import get_root_path
from werkzeug.datastructures import ImmutableDict
from werkzeug.urls import url_parse

from indico.util.caching import make_hashable
from indico.util.fs import resolve_link
from indico.util.packaging import package_is_editable
from indico.util.string import crc32, snakify


DEFAULTS = {
    'ATTACHMENT_STORAGE': 'default',
    'AUTH_PROVIDERS': {},
    'BASE_URL': None,
    'CACHE_BACKEND': 'files',
    'CACHE_DIR': '/opt/indico/cache',
    'CATEGORY_CLEANUP': {},
    'CELERY_BROKER': None,
    'CELERY_CONFIG': {},
    'CELERY_RESULT_BACKEND': None,
    'COMMUNITY_HUB_URL': 'https://hub.getindico.io',
    'CUSTOMIZATION_DEBUG': False,
    'CUSTOMIZATION_DIR': None,
    'CUSTOM_COUNTRIES': {},
    'CUSTOM_LANGUAGES': {},
    'DB_LOG': False,
    'DEBUG': False,
    'DEFAULT_LOCALE': 'en_GB',
    'DEFAULT_TIMEZONE': 'UTC',
    'DISABLE_CELERY_CHECK': None,
    'ENABLE_ROOMBOOKING': False,
    'EXPERIMENTAL_EDITING_SERVICE': False,
    'EXTERNAL_REGISTRATION_URL': None,
    'HELP_URL': 'https://learn.getindico.io',
    'IDENTITY_PROVIDERS': {},
    'LOCAL_IDENTITIES': True,
    'LOCAL_MODERATION': False,
    'LOCAL_REGISTRATION': True,
    'LOCAL_GROUPS': True,
    'LOGGING_CONFIG_FILE': 'logging.yaml',
    'LOGO_URL': None,
    'LOG_DIR': '/opt/indico/log',
    'MAX_UPLOAD_FILES_TOTAL_SIZE': 0,
    'MAX_UPLOAD_FILE_SIZE': 0,
    'MEMCACHED_SERVERS': [],
    'NO_REPLY_EMAIL': None,
    'PLUGINS': set(),
    'PROFILE': False,
    'PROVIDER_MAP': {},
    'PUBLIC_SUPPORT_EMAIL': None,
    'REDIS_CACHE_URL': None,
    'ROUTE_OLD_URLS': False,
    'SCHEDULED_TASK_OVERRIDE': {},
    'SECRET_KEY': None,
    'SENTRY_DSN': None,
    'SENTRY_LOGGING_LEVEL': 'WARNING',
    'SESSION_LIFETIME': 86400 * 31,
    'SMTP_LOGIN': None,
    'SMTP_PASSWORD': None,
    'SMTP_SERVER': ('localhost', 25),
    'SMTP_TIMEOUT': 30,
    'SMTP_USE_CELERY': True,
    'SMTP_USE_TLS': False,
    'SQLALCHEMY_DATABASE_URI': None,
    'SQLALCHEMY_MAX_OVERFLOW': 3,
    'SQLALCHEMY_POOL_RECYCLE': 120,
    'SQLALCHEMY_POOL_SIZE': 5,
    'SQLALCHEMY_POOL_TIMEOUT': 10,
    'STATIC_FILE_METHOD': None,
    'STATIC_SITE_STORAGE': None,
    'STORAGE_BACKENDS': {'default': 'fs:/opt/indico/archive'},
    'STRICT_LATEX': False,
    'SUPPORT_EMAIL': None,
    'TEMP_DIR': '/opt/indico/tmp',
    'USE_PROXY': False,
    'WORKER_NAME': socket.getfqdn(),
    'XELATEX_PATH': None,
}

# Default values for settings that cannot be set in the config file
INTERNAL_DEFAULTS = {
    'CONFIG_PATH': os.devnull,
    'CONFIG_PATH_RESOLVED': None,
    'LOGGING_CONFIG_PATH': None,
    'TESTING': False
}


def get_config_path():
    """Get the path of the indico config file.

    This may return the location of a symlink.  Resolving a link is up
    to the caller if needed.
    """
    # In certain environments (debian+uwsgi+no-systemd) Indico may run
    # with an incorrect $HOME (such as /root), resulting in the config
    # files being searched in the wrong place. By clearing $HOME, Python
    # will get the home dir from passwd which has the correct path.
    old_home = os.environ.pop('HOME', None)
    # env var has priority
    try:
        return os.path.expanduser(os.environ['INDICO_CONFIG'])
    except KeyError:
        pass
    # try finding the config in various common paths
    paths = [os.path.expanduser('~/.indico.conf'), '/etc/indico.conf']
    # Keeping HOME unset wouldn't be too bad but let's not have weird side-effects
    if old_home is not None:
        os.environ['HOME'] = old_home
    # If it's an editable setup (ie usually a dev instance) allow having
    # the config in the package's root path
    if package_is_editable('indico'):
        paths.insert(0, os.path.normpath(os.path.join(get_root_path('indico'), 'indico.conf')))
    for path in paths:
        if os.path.exists(path):
            return path
    raise Exception('No indico config found. Point the INDICO_CONFIG env var to your config file or '
                    'move/symlink the config in one of the following locations: {}'.format(', '.join(paths)))


def _parse_config(path):
    globals_ = {'timedelta': timedelta, 'crontab': crontab}
    locals_ = {}
    with codecs.open(path, encoding='utf-8') as config_file:
        # XXX: unicode_literals is inherited from this file
        exec(compile(config_file.read(), path, 'exec'), globals_, locals_)
    return {str(k if k.isupper() else _convert_key(k)): v
            for k, v in locals_.items()
            if k[0] != '_'}


def _convert_key(name):
    # camelCase to BIG_SNAKE while preserving acronyms, i.e.
    # FooBARs -> FOO_BARS (and not FOO_BA_RS)
    name = re.sub(r'([A-Z])([A-Z]+)', lambda m: m.group(1) + m.group(2).lower(), name)
    name = snakify(name).upper()
    special_cases = {'PDFLATEX_PROGRAM': 'XELATEX_PATH',
                     'IS_ROOM_BOOKING_ACTIVE': 'ENABLE_ROOMBOOKING'}
    return special_cases.get(name, name)


def _postprocess_config(data):
    if data['DEFAULT_TIMEZONE'] not in pytz.all_timezones_set:
        raise ValueError('Invalid default timezone: {}'.format(data['DEFAULT_TIMEZONE']))
    data['BASE_URL'] = data['BASE_URL'].rstrip('/')
    data['STATIC_SITE_STORAGE'] = data['STATIC_SITE_STORAGE'] or data['ATTACHMENT_STORAGE']
    if data['DISABLE_CELERY_CHECK'] is None:
        data['DISABLE_CELERY_CHECK'] = data['DEBUG']


def _sanitize_data(data, allow_internal=False):
    allowed = set(DEFAULTS)
    if allow_internal:
        allowed |= set(INTERNAL_DEFAULTS)
    for key in set(data) - allowed:
        warnings.warn(f'Ignoring unknown config key {key}')
    return {k: v for k, v in data.items() if k in allowed}


def load_config(only_defaults=False, override=None):
    """Load the configuration data.

    :param only_defaults: Whether to load only the default options,
                          ignoring any user-specified config file
                          or environment-based overrides.
    :param override: An optional dict with extra values to add to
                     the configuration.  Any values provided here
                     will override values from the config file.
    """
    data = dict(DEFAULTS, **INTERNAL_DEFAULTS)
    if not only_defaults:
        path = get_config_path()
        config = _sanitize_data(_parse_config(path))
        data.update(config)
        env_override = os.environ.get('INDICO_CONF_OVERRIDE')
        if env_override:
            data.update(_sanitize_data(ast.literal_eval(env_override)))
        resolved_path = resolve_link(path) if os.path.islink(path) else path
        resolved_path = None if resolved_path == os.devnull else resolved_path
        data['CONFIG_PATH'] = path
        data['CONFIG_PATH_RESOLVED'] = resolved_path
        if resolved_path is not None:
            data['LOGGING_CONFIG_PATH'] = os.path.join(os.path.dirname(resolved_path), data['LOGGING_CONFIG_FILE'])

    if override:
        data.update(_sanitize_data(override, allow_internal=True))
    _postprocess_config(data)
    return ImmutableDict(data)


class IndicoConfig:
    """Wrapper for the Indico configuration.

    It exposes all config keys as read-only attributes.

    Dynamic configuration attributes whose value may change depending
    on other factors may be added as properties, but this should be
    kept to a minimum and is mostly there for legacy reasons.

    :param config: The dict containing the configuration data.
                   If omitted, it is taken from the active flask
                   application.  An explicit configuration dict should
                   not be specified except in special cases such as
                   the initial app configuration where no app context
                   is available yet.
    :param exc: The exception to raise when accessing an invalid
                config key.  This allows using the expected kind of
                exception in most cases but overriding it when
                exposing settings to Jinja where the default
                :exc:`AttributeError` would silently be turned into
                an empty string.
    """

    __slots__ = ('_config', '_exc')

    def __init__(self, config=None, exc=AttributeError):
        # yuck, but we don't allow writing to attributes directly
        object.__setattr__(self, '_config', config)
        object.__setattr__(self, '_exc', exc)

    @property
    def data(self):
        try:
            return self._config or current_app.config['INDICO']
        except KeyError:
            raise RuntimeError('config not loaded')

    @property
    def hash(self):
        return crc32(repr(make_hashable(sorted(self.data.items()))))

    @property
    def CONFERENCE_CSS_TEMPLATES_BASE_URL(self):
        return self.BASE_URL + '/css/confTemplates'

    @property
    def IMAGES_BASE_URL(self):
        return 'static/images' if g.get('static_site') else url_parse(f'{self.BASE_URL}/images').path

    @property
    def LATEX_ENABLED(self):
        return bool(self.XELATEX_PATH)

    def __getattr__(self, name):
        try:
            return self.data[name]
        except KeyError:
            raise self._exc('no such setting: ' + name)

    def __setattr__(self, key, value):
        raise AttributeError('cannot change config at runtime')

    def __delattr__(self, key):
        raise AttributeError('cannot change config at runtime')


#: The global Indico configuration
config = IndicoConfig()
